src/client/sound_openal.cpp

```diff
@@ -616,38 +614,40 @@ class OpenALSoundManager: public ISoundManager
 
 	void fadeSound(int soundid, float step, float gain)
 	{
-		m_sounds_fading[soundid] = FadeState(step, getSoundGain(soundid), gain);
+		if (step == 0)
+			return;
+		if (m_sounds_fading.find(soundid) != m_sounds_fading.end()) {
+			if(m_sounds_fading[soundid].target_gain == gain)
+				return;
+			m_sounds_fading.erase(soundid);
+		}
+		float current_gain = getSoundGain(soundid);
+		gain = rangelim(gain, 0, 1);
+		step = gain - current_gain > 0 ? abs(step) : -abs(step);
+		m_sounds_fading[soundid] = FadeState(step, current_gain, gain);
 	}
 
 	void doFades(float dtime)
 	{
-		m_fade_delay += dtime;
-
-		if (m_fade_delay < 0.1f)
-			return;
+		for (auto i = m_sounds_fading.begin(); i != m_sounds_fading.end();) {
+			i->second.current_gain += (i->second.step * dtime);
```

Comment:
A `FadeState &s = i->second;` before here would improve the readability of the rest